hugo/content/docs/ast-types.md

```diff
@@ -3,4 +3,340 @@ title: "AST Type Inference"
+
+Let's look at the example from the previous section:
+```
+X infers MyType: name=ID;
+Y infers MyType: name=ID count=INT;
+
+// should be replaced by:
+interface MyType {
+    name: string
+    count?: number
+}
+
+X returns MyType: name=ID;
+Y returns MyType: name=ID count=INT;
+```
+We now explicitly declare `MyType` directly in the grammar with the keyword `interface`. The parser rules `X` and `Y` creating AST nodes of type `MyType` need to explicitly declare the type of the AST node they create with the keyword `returns`.
+
+Contrary to [inferred types](#inferred-types), all properties must be explicitly declared in order to be valid inside of a parser rule. The following syntax:
+```
+Z returns MyType: name=ID age=INT;
+```
+will throw the following validation error `A property 'age' is not expected` because the declaration of `MyType` does not include the property `age`. In short, *declared types* add a layer of safety via validation to the grammar that prevents mismatches between the generated AST types and what the user expects.
+
+A declared type can also extend another declared type:
+```
+interface MyType {
+    name: string
+}
+
+interface MyOtherType extends MyType {
+    count: number
+}
+
+Y returns MyOtherType: name=ID count=INT;
+```
+
+Explicitly declaring type aliases in the grammar is achieved with the keyword `type`:
+```
+type X = A | B;
+
+// generates:
+type X = A | B;
+```
+
+Please note that it is not allowed to use an alias type as a return type in a parser rule. The following syntax is invalid:
+```
+type X = A | B;
+
+Y returns X: name=ID;
+```
+
+Using `returns` without explicitly declaring the type of the AST node is not allowed and results in a validation error.
+
+### Cross-references, Arrays, and Alternatives
+Declared types come with special syntax to declare cross-references, arrays, and alternatives:
+```
+interface A {
+    reference: @B
+    array: B[]
+    alternative: B | C
+}
+
+interface B {
+    name: string
+}
+
+interface C {
+    name: string
+    count: number
+}
+
+X returns A: reference=[B:ID] array+=Y (array+=Y)* alternative=(Y | Z);
+
+Y returns B: 'Y' name=ID;
+
+Z returns C: 'Z' name=ID count=INT;
+```
+
+### Actions
+Actions referring to a declared type have the following syntax:
+```
+interface A {
+    name: string
+}
+
+interface B {
+    name: string
+    count: number
+}
+
+X: 
+    {A} 'A' name=ID 
+  | {B} 'B' name=ID count=INT;
+```
+Note the absence of the keyword `infer`, contrary to [actions inferring types](#simple-actions).
+
+## Refactoring Dummy Rules
```

Comment:
I'm not too sure about this section in general. It sounds really forced in a way and is probably only for people coming from Xtext? If that's the case, we should make the reference here explicit, and if it's not, then we should probably remove this section.

Reply:
I'll leave it there for now and maybe we should think about having a documentation page about migrating from XText.